### PR TITLE
[goodgame] Add goodgame extractors

### PR DESCRIPTION
**Pull request by tausackhn**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Add [goodgame.ru](goodgame.ru) extractor for streams, videos, clips. Site support request #9058.

### CODE REVIEW

File: youtube_dl/extractor/goodgame.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        matches = list(re.finditer(r'<div[^>]+class=\"title\"[^>]*>(?P<title>[^<]+)', webpage))
+        title = matches[1].group('title')
+        description = self._html_search_regex(r'<div[^>]+class=\"description\"[^>]*>(?P<info>[^\0]*?)</div>',
+                                              webpage, 'info')
```
Comment (dstftw) [inline youtube_dl/extractor/goodgame.py]: Read coding conventions and fix all optional meta fields.

File: youtube_dl/extractor/goodgame.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        timestamp = int_or_none(self._html_search_regex(self._RE_TIMESTAMP, webpage, 'timestamp'))
+        uploader_match = re.search(self._RE_UPLOADER, webpage)
+        uploader = uploader_match.group('uploader')
+        uploader_id = uploader_match.group('uploader_id')
+
+        formats = [{
+            'format_id': 'clip',
+            'url': url,
+            # http or https extraction
+            'protocol': url[:6].rstrip(':/'),
```
Comment (dstftw) [inline youtube_dl/extractor/goodgame.py]: No need to specify this.

File: youtube_dl/extractor/goodgame.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        'only_matching': True,
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._og_search_title(webpage)
+        thumbnail = self._og_search_thumbnail(webpage)
+        url = self._og_search_video_url(webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/goodgame.py]: Don't shadow method argument name.

File: youtube_dl/extractor/goodgame.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            }
+
+        file = self._html_search_regex(r'<param[^>]+name=\"flashvars\"[^>]+value=\"src=(?P<path>[^=\"]+)\"[^>]*>',
+                                       webpage, 'path', default=None)
+        if file:
+            rtmp_url = '%s%s' % (self._RTMP_SERVER, file)
+            formats = [{
+                'format_id': 'rtmp',
+                'url': rtmp_url,
+                'protocol': 'rtmp',
```
Comment (dstftw) [inline youtube_dl/extractor/goodgame.py]: No need to specify this.

File: youtube_dl/extractor/goodgame.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+                'ext': 'mp4',
```
Comment (dstftw) [inline youtube_dl/extractor/goodgame.py]: Should be `flv`.

File: youtube_dl/extractor/goodgame.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+        description = self._html_search_regex(r'<div[^>]+class=\"description\"[^>]*>(?P<info>[^\0]*?)</div>',
+                                              webpage, 'info')
+        timestamp = int_or_none(self._html_search_regex(self._RE_TIMESTAMP, webpage, 'timestamp'))
+        uploader_match = re.search(self._RE_UPLOADER, webpage)
+        uploader = uploader_match.group('uploader')
+        uploader_id = uploader_match.group('uploader_id')
+
+        embed_url = self._html_search_regex(
+            r'<iframe[^>]+src="((?:https?:)?//(?:www\.)?youtube\.com/embed[^"]+)"',
+            webpage, 'embed URL', default=None)
```
Comment (dstftw) [inline youtube_dl/extractor/goodgame.py]: `YoutubeIE._extract_url`.

**Comment by tausackhn** [comment]
Hi, Sergey! Are any other corrections needed?

**Comment by tri4ng1e** [comment]
Not working if "The broadcaster indicated that the channel is intended for mature audiences", it fails to get id.
```
Traceback (most recent call last):
  File "./youtube-dl/youtube_dl/YoutubeDL.py", line 784, in extract_info
    ie_result = ie.extract(url)
  File "./youtube-dl/youtube_dl/extractor/common.py", line 434, in extract
    ie_result = self._real_extract(url)
  File "./youtube-dl/youtube_dl/extractor/goodgame.py", line 90, in _real_extract
    _id = self._search_regex('src=\"https://goodgame.ru/player\?(?P<id>\d+)\"', stream_info.get('embed'), 'id')
  File "./youtube-dl/youtube_dl/extractor/common.py", line 797, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name)
```
Also when stopping the record and then starting it again, it says that file already downloaded. So I think that something like timestamp in file name (like `streamer-name_1970-03-26 23:59:59`) will be nice.

**Comment by tausackhn** [comment]
Could you give an example of url? I've tested for random channel with such notice and all works fine. Youtube-dl chooses best quality by default so it is `Source`.

**Comment by tri4ng1e** [comment]
I can't find that streamer now, it was random one, will ask friends, they have them. I found one stream with simmilar thing, but looks like it's not the only one setting for mature content. Earlier i had like "broadcaster indicated" warning, and now I found like "This broadcast may contain an obscene speech...", and in that case all worked.

Yes, it is. But. Random streamer https://goodgame.ru/channel/twaryna/
```
./youtube-dl https://goodgame.ru/channel/twaryna/
```
default quality

```
./youtube-dl https://goodgame.ru/channel/twaryna/ -f "Source"
```
source quality

(Sorry for my bad English)

**Comment by tausackhn** [comment]
@tri4ng1e i've fixed format preferences. You should specify [output template](https://github.com/rg3/youtube-dl#output-template) if you want to add timestamps or something like that, e.g. `epoch`, `autonumber`.

**Comment by soredake** [comment]
any progress on this?

**Pull request closed without merging**